### ISSUE TITLE
Release 0.2.1?

### ISSUE THREAD

**Issue by sammyke007** [stas-demydiuk/domoticz-zigbee2mqtt-plugin#549]
Hi

Tnx for your work. Untill v0.2.1 this plugin has been working flawless for me.
The main code page still mentions v0.2.1, but this should be v3.0 now, correct?

Is it safe now to update on a running system? I know I will have to remove all of my Z2M devices in Domoticz and re-add them again (with losing all history/logs), but if this is the way, than I will update and bite the apple ;-).

**Comment by Anjerlaan** [comment]
@sammyke007 which version of Z2M are you running?

**Comment by sammyke007** [comment]
1.17.0, master

**Comment by Anjerlaan** [comment]
FYI In this version the new_api is mandatory, so in configuration.yaml there is no need to add:

```
experimental:
new_api: true
```

After you set this, remove the domoticz-zigbee2mqtt plugin from the hardware page.
Update to version V3
And add the plug-in adapter in the hardware page again.

Is your current setup working with Z2M 1.17.0, do you have errors?

**Comment by sammyke007** [comment]
But removing the domoticz-zigbee2mqtt plugin from the hardware page means losing all devices in Domoticz and having to readd them all again? With losing all your logs/history?

For now ZM 1.17.0 with previous plugin 0.2.1 is working great, but I know I will have to update anyway soon...

**Comment by hughb8on** [comment]
If you remove the plugin from the hardware page you will lose all your zigbee devices, yes.

This is a show-stopper for me. 

I would lose so much work it just wouldn't be worth it. As well as the history I would lose On/Off actions, scripts, notifications, timers and my Alexa integration would need considerable effort to fix it. Believe me, I know, I've tried v3.0 on a test system.

Even if it wasn't necessary to trash-and-recreate everything, the plugin's creation of existing devices with new names and its orphaning of old ones would still make a move very problematic. Also, the policy of creating a separate domoticz device for every attribute exposed by zigbee2mqtt.io is really asking for trouble because of the 255 device restriction imposed by the framework.

I really like 0.2.1 and greatly appreciate the tremendous work which went into it. But because of the problems mentioned above I really have no choice at this point but to stick with (my own fork of) it, adding the bits I want as I go.

**Comment by heggink** [comment]
Can I just check my understanding: if I upgrade z2m (from my 1.6.2 version to 1.17) then that will keep all the devices. If I then delete the current plugin and install the new one, wouldn't the plugin create new devices for every device that z2m exposes to the plugin? Or do I really need to reinclude all devices again?

**Comment by hughb8on** [comment]
If you just upgrade 1.16.2 to 1.17 that's not a problem, your devices will remain available in domoticz and no change to the plugin becomes necessary just because you do that. 

However, if you do decide to upgrade the plugin and follow the instructions to delete it first, then all its devices will be deleted from domoticz.

Yes, they will be recreated when you install the new plugin. Some will have slightly different names and some new ones will appear too. That may not be a problem to you - it won't be a big deal to people who don't have as many devices as I have, and/or have fewer of the dependencies I listed above.

If that fits you, then I'd encourage you to upgrade.

**Comment by sammyke007** [comment]
I'm staying away from updating for now...

**Comment by Anjerlaan** [comment]
> If you just upgrade 1.16.2 to 1.17 that's not a problem, your devices will remain available in domoticz and no change to the plugin becomes necessary just because you do that.
> 
> However, if you do decide to upgrade the plugin, and you follow the instructions to delete it first, then all its devices will be deleted from domoticz.
> 
> Yes, they will be automatically recreated when you install the new plugin. No re-pairing or anything is required. Some of the 'new' domoticz devices will have slightly different names and some additional ones will appear too. That may not be a problem to you - it won't be a big deal to people who don't have as many devices as I have, and/or have fewer of the dependencies I listed above.
> 
> If that fits you, then I'd encourage you to upgrade.

Is in the official release 0.2.1 the listing of the ZigBee devices problem solved when using Z2m >1.16.2? https://github.com/stas-demydiuk/domoticz-zigbee2mqtt-plugin/issues/458#issuecomment-744439306
I know it was updated at a certain point, but I am not sure if the official 0.2.1 release also has this change?
If not could this be a problem when stepping back to 0.2.1 release?

In the Settings of the v3.0 plugin you can turn of the battery devices and tracking of link quality. If also the effect_devices are added to these options (I think you mentioned that?), that would decrease the amount of devices significantly isn't it?

**Comment by heggink** [comment]
I can't update until the new firmware for my radiator thermostat is released as I have a working workaround that I don;t want to recreate.
Luckily, most of my scripts are device name based so if these show up again, most will work but I will have to re-add to groups and scenes which is quite some work. 
That said, I know that sometimes breaking changes are difficult to prevent. The switch to the new API makes total sense to me so will have to bite the bullet at some point. I guess the devices being recreated with the new plugin is the closest to an autodiscovery feature taht I would really love (similar with zwave which is much more problematic and the reason for me to adopt zigbee).
Thanks for the feedback!

**Comment by Anjerlaan** [comment]
> most of my scripts are device name based so if these show up again, most will work but I will have to re-add to groups and scenes which is quite some work.

@heggink I was just thinking...Isn't it possible to create a (dzVents) script which automatically creates (your) Groups and Scenes and add the devices and their appropriate settings? Since most of your scripts use the device name and not the idx, this script would always be usable if a rebuild is needed?

here is some information https://www.domoticz.com/forum/viewtopic.php?p=266456#p266456

**Comment by hughb8on** [comment]
> Is in the official release 0.2.1 the listing of the ZigBee devices problem solved when using Z2m >1.16.2? [#458 (comment)](https://github.com/stas-demydiuk/domoticz-zigbee2mqtt-plugin/issues/458#issuecomment-744439306)
> I know it was updated at a certain point, but I am not sure if the official 0.2.1 release also has this change?
> If not could this be a problem when stepping back to 0.2.1 release?

I believe this problem was fixed by Koenkk regressing his change which caused problems for those on 0.2.1 plus 1.16.2.

I can confirm that, as sammykee007 says above, 0.2.1 works fine with 1.17.0

**Comment by hughb8on** [comment]
> In the Settings of the v3.0 plugin you can turn of the battery devices and tracking of link quality. If also the effect_devices are added to these options (I think you mentioned that?), that would decrease the amount of devices significantly isn't it?

It would help, but v3.0 now adds sub-devices pretty much in line with whatever zigbee2mqtt.io exposes. For instance it now creates "power" devices (to track/amend power-on behaviour).

What would be good would be a way of selecting exactly which sub-devices you want , on a per-device basis. 

Or alternatively for the plugin to "remember" which of the ones it created you then deleted. At the moment of course it just recreates them.

**Comment by hughb8on** [comment]
I wish I had the skill to actually make these changes rather than just suggest the need for them. As I work on my 0.2.1 fork I may acquire some. Sorry @stas-demydiuk!

**Comment by stas-demydiuk** [comment]
> What would be good would be a way of selecting exactly which sub-devices you want , on a per-device basis.

I also would like to have such a feature but unfortunately, Domoticz's plugin system allows us to have only up to 6 settings per plugin and we already use 5 of them. So there is currently no space to add such a feature

**Comment by hughb8on** [comment]
I see, yes. I also understand and support the move to the new api.

May I suggest a possible way forward on the sub-devices problem? 

Perhaps the plugin should check whether there is already at least one domoticz device with that IEEE Address, and not create any "missing" sub-devices if it finds one.

That would mean that sub-devices could be deleted manually by those who don't want them and the plugin wouldn't recreate them.

It would also mean that those wanting new sub-devices, or perhaps who had changed their minds / accidentally deleted sub-devices could simply delete all the domoticz devices for that IEEE address and the plugin would create a fresh full set.

And it would remove the need to use any of those six domoticz settings, freeing them for future use.

**Comment by Anjerlaan** [comment]
> > What would be good would be a way of selecting exactly which sub-devices you want , on a per-device basis.
> 
> I also would like to have such a feature but unfortunately, Domoticz's plugin system allows us to have only up to 6 settings per plugin and we already use 5 of them. So there is currently no space to add such a feature

Isn't it possible to extract the MQTT credentials (Login name, password and Client ID) from Zigbee2Mqtt configuration.yaml?
This would save 3 settings

**Comment by hughb8on** [comment]
If that file is accessible; the problem is the MQTT instance may not be running on the same system.

**Comment by Anjerlaan** [comment]
> If that file is accessible; the problem is the MQTT instance may not be running on the same system.

The Server address and port are still settings in the domoticz z2m plugin which need to be filled in.....this tells you where mosquitto is located isn't it?....if so, then Z2M is also located there....then the path to the configuration.yaml is also known.

**Comment by hughb8on** [comment]
I do see where you are coming from but I disagree with this approach for several reasons. 

Firstly, the the path to the configuration file is *not* known. You definitely don't know the path to mine.

Secondly, and more importantly, if somebody goes to the trouble of requiring credentials to access an MQTT broker then it *should* be hard for a random user on a remote system to extract them. The minimum requirement would be that accessing them would require authentication at least as strict as that imposed by the broker. The plugin would have to store the credentials required to access the MQTT credentials, which would leave us no better off.

**Comment by Anjerlaan** [comment]
You're right, the plugin is only publishing Z2M messages to the mqtt broker, it doesn't know where Z2M is installed.

**Comment by Anjerlaan** [comment]
> I can't update until the new firmware for my radiator thermostat is released as I have a working workaround that I don;t want to recreate.
> Luckily, most of my scripts are device name based so if these show up again, most will work but I will have to re-add to groups and scenes which is quite some work.
> That said, I know that sometimes breaking changes are difficult to prevent. The switch to the new API makes total sense to me so will have to bite the bullet at some point. I guess the devices being recreated with the new plugin is the closest to an autodiscovery feature taht I would really love (similar with zwave which is much more problematic and the reason for me to adopt zigbee).
> Thanks for the feedback!

You could use this bash script to update all group and scene devices.
This is updating the scene device names with the new idx number.
Read the thread

https://www.domoticz.com/forum/viewtopic.php?f=59&t=35244&p=266955#p266955


```
backupDatabase=/home/pi/domoticz/backups/daily/backup-day-12-domoticz.db # change to real name of backup
currentDatabase=/home/pi/domoticz/domoticz.db  # change to real name of current database

sudo service domoticz stop
sudo cp $currentDatabase safeCopy.db # just to be sure
sudo cp $currentDatabase dz.db

sudo sqlite3 dz.db "ATTACH DATABASE \"$backupDatabase\" AS backup;
CREATE TABLE ConversionHelper AS SELECT * FROM backup.devicestatus;
CREATE VIEW Conversion AS SELECT a.ID New, b.ID Old, b.Name Name FROM DeviceStatus a, ConversionHelper b WHERE a.Name = b.Name;
UPDATE SceneDevices SET DeviceRowID = (SELECT c.New FROM conversion c WHERE c.Old = DeviceRowID );
DROP VIEW Conversion;
DROP TABLE ConversionHelper;"

sudo mv dz.db domoticz_converted.db

echo Scene devices converted
echo Check database and if OK replace current domoticz database with domoticz_converted.db
echo and start domoticz 
```

**Comment by heggink** [comment]
@Anjerlaan  thanks! will definitely consider once the firmware of the eurotronics thermostat is updated and working. I don't have a ton of groups/scenes using these devices but still useful (I have traditionally been very caustious doing sqlite db edits :-O).

**Comment by hughb8on** [comment]
I have worked out, tested, and successfully implemented a method for upgrading from 0.2.1 to v3.0 which doesn't require uninstalling the plugin and thereby deleting all zigbee devices.

To cut a long story short, I manually updated domoticz.db to change the Device IDs to match the new names which v3.0 uses.

Using a test domoticz instance I noted the IDs which change and prepared a list of SQL statements of the form: -

UPDATE devicestatus SET deviceid = '0x000b3cfffef4febd_state' WHERE deviceid = '0x000b3cfffef4febd_switch';

In my case there were 28 such name changes.

The result is that all of my zigbee devices have been preserved with their original IDXs, and hence all of the stuff listed in my post above continues to work without any further intervention.

**Comment by sammyke007** [comment]
I'm interested!

**Comment by sammyke007** [comment]
OK... I've finally decided to bite the bullet and upgraded from 0.2.1 to 3.0.
I removed the old plugin ( Setup > Hardware ). This deleted all of my Z2M devices in Domoticz.
I removed the zigbee2mqtt folder completely and made a fresh pull.
All my Z2M devices got added again (with new IDX's) after adding the hardware again. I had to rescript and reconfigure some things, but all in all, glad I've done it and it' working. Tnx stas-demydiuk!

PS: maybe update the release to 3.0 instead of 0.2.1 on the front page? It's still confusing.
